Add changelog entry for component's configuration (#2928)
alphagov/govuk-frontend: CHANGELOG.md

This was added in [#2911 Add warning suppression functionality](https://github.com/alphagov/govuk-frontend/pull/2911)

#### Configure components in JavaScript

> JavaScript components receiving configuration through data attributes now accept the same configuration options when instantiated in JavaScript. These components are:

Reviewer: Is there a plainer option for 'instantiated' that is still accurate? Would something like this be accurate in any way? - ' ...the same configuration options as when an instance is created in JavaScript.'

Reviewer: Or maybe we can turn this sentence around a bit to put a different emphasise? 
'JavaScript components can now get the same configuration options in 2 ways - through data attributes and when creating an instance.'

Author: That second sentence feels more direct, I like it 😄 It muddies the fact that the data attributes were already there and that the change is about receiving configuration in JavaScript, though. Would it hurt to add some timing precisions in the second part:

> JavaScript components can ~~now~~ get the same configuration options in 2 ways - through data attributes, **as before,** and **now** when creating an instance.

Reviewer: Yup, that works. Fab 😄

- the `Button` component, for its `preventDoubleClick` option (matching `data-prevent-double-click`)
- the `CharacterCount` component, for its `maxlength`,`maxwords` and `threshold` options (matching `data-maxlength`, `data-maxwords` and `data-threshold`, respectively)
- the `ErrorSummary` component, for its `disableAutoFocus` option (matching `data-disable-auto-focus`)
- the `NotificationBanner` component, for its `disableAutoFocus` option (matching `data-disable-auto-focus`)

> You can omit these options when calling the Nunjucks macro and provide them:

Reviewer: ```suggestion
You can omit these configuration options when calling the Nunjucks macro and provide them:
```

- at component instantiation, in a configuration object as second argument, for example:

>   ```js

Reviewer: It would be better if the examples were not part of the bullet list.

Author: Oh, I thought it would help to keep them tied to the situation they give an example for. Is it because they break the flow when reading through the list? 

I've pushed an [updated version addressing this and the previous comment](https://github.com/alphagov/govuk-frontend/blob/8cd406c51964f988557bbee47513dc432e42d170/CHANGELOG.md#configure-components-in-javascript). (Here's the [old version for comparison](https://github.com/alphagov/govuk-frontend/blob/ac8e7820f3150452bf2ff2c50a5c20877db1cefb/CHANGELOG.md#configure-components-in-javascript)).

> ```js
> var errorSummary = document.querySelector('[data-module="error-summary"]')
> new GOVUKFrontend.ErrorSummary(errorSummary, {disableAutoFocus: true})
> ```

Reviewer 2: Suggest indenting this so that it falls within the list item:

```suggestion
  ```js
  var errorSummary = document.querySelector('[data-module="error-summary"]')
  new GOVUKFrontend.ErrorSummary(errorSummary, {disableAutoFocus: true})
  ```
```

Reviewer 2: `disableAutoFocus` is going to be irrelevant for most if not all service teams – we only really have it for use in examples in the Design System.

It might be worth providing examples for Button with the `preventDoubleClick` option instead?

- when initialising components in bulk using `initAll`, for example:

```js
GOVUKFrontend.initAll({
  errorSummary: {
    disableAutoFocus: true
  }
  // Or, for the other components,
  // button: {/* options */},
  // characterCount: {/* options */},
  // notificationBanner: {/* options */}
})
```

> If you provide an option both through data attribute and JavaScript configuration, the components will use the value from the data attribute.

Reviewer 2: We might be able to hand off to the frontend docs for this level of detail

(It'd make sense to include a link either way)

> This was added in pull requests specific for each components: [NotificationBanner – #2843](https://github.com/alphagov/govuk-frontend/pull/2843), [ErrorSumarry – #2854](https://github.com/alphagov/govuk-frontend/pull/2854), [Button – #2883](https://github.com/alphagov/govuk-frontend/pull/2883), and [CharacterCount – #2887](https://github.com/alphagov/govuk-frontend/pull/2887)

Reviewer 2: These PR numbers don't look quite right – #2883 is listed for the Button but is actually the PR for Character Count and #2887 is the Character Count i18n work.

I'd also suggest using the 'human' component names (i.e. add a space so 'Notification Banner')

Author: Argh, sligth "off by one copy paste" followed by "didn't double check the links", sorry 😔

This was added in [#2911 Add warning suppression functionality](https://github.com/alphagov/govuk-frontend/pull/2911)

#### Configure components in JavaScript

JavaScript components can get the same configuration options in 2 ways - through data attributes, as before, and now when creating an instance. These components are:

- the `Button` component, for its `preventDoubleClick` option (matching `data-prevent-double-click`)
- the `CharacterCount` component, for its `maxlength`,`maxwords` and `threshold` options (matching `data-maxlength`, `data-maxwords` and `data-threshold`, respectively)
- the `ErrorSummary` component, for its `disableAutoFocus` option (matching `data-disable-auto-focus`)
- the `NotificationBanner` component, for its `disableAutoFocus` option (matching `data-disable-auto-focus`)

> You can omit these configuration options when calling the Nunjucks macro and provide them:

Reviewer: ```suggestion
You can leave out these configuration options when using the Nunjucks macro and provide configuration when:
```

> - at component instantiation, in a configuration object as second argument

Reviewer: ```suggestion
- creating a component, in a configuration object as second argument
```

> - when initialising components in bulk using `initAll`

Reviewer: ```suggestion
- initialising components in bulk using `initAll`
```

Reviewer: added some minor suggestions for clarity. 
Just checking whether there are any acknowledgements to contributors to add?

Author: @claireashworth I'm not sure what you mean by "acknowledgements to contributors to add". Do you mean "
acknowledging external contributors? If so, we're good for that entry as it was Ollie's and mine doing 😄

Reviewer: 👍🏻 That's exactly what I meant. Just checking for when we make release notes as well.
